### PR TITLE
Add optimization to deduplicate constants

### LINKED ISSUE #1795: Optimization to deduplicate literals within dask graphs
We would like to sometimes rewrite graphs like the following:

```python
x = np.ones(1000000)

dsk = {'a': (add, x, 1),
       'b': (sub, x, 1)}
```

to

```python
dsk = {'x': x,
      'a': (add, 'x', 1),
      'b': (sub, 'x', 1)}
```

This is a bit like the opposite operation of inlining.  Note that we pulled out the literal `x`, but not `1`.  We might want to pull out objects of a particular type or that match a particular predicate, such as being expensive to serialize.

cc @eriknw , in case you're interested.

### PR DESCRIPTION
**Pull request by jcrist**
Adds a function ``dedupe_constants`` to remove duplicate constants in
dask graphs. The function takes a dask graph and a predicate function
indicating whether a constant may be extracted. Any duplicate
extractable constants are pulled out into their own key, with the key
name determined by ``tokenize``.

Example:

```python
>>> big_str = 'abcde' * 1000
>>> d = {'a': (len, big_str), 'b': (len, big_str), 'c': (add, 'a', 'b')}
>>> def is_big_str(x):
...     return isinstance(x, str) and len(x) > 1000
>>> dsk = dedupe_constants(d, is_big_str)
>>> dsk
{'a': (len, 'const-ee82b7de828a74a003c1228ff7e186cf'),
 'b': (len, 'const-ee82b7de828a74a003c1228ff7e186cf'),
 'c': (add, 'a', 'b'),
 'const-ee82b7de828a74a003c1228ff7e186cf': 'abcdeabcdeab...'}
```

Fixes #1795.

### CODE REVIEW

**Comment by pitrou** [comment]
Now that I think about it, I wonder about the point of this optimization. `pickle` is smart and won't serialize the same object twice: if it meets two references to the same object, the second reference will be serialized as a simple index to the first one (in technical terms, you can look at the `memo` dictionary in pickle's implementation).

So, unless the graph is serialized piecewise with distinct calls to `dumps()`, this shouldn't have any positive impact.

**Comment by pitrou** [comment]
Oh, I understand: this is about reducing overhead of serializing from the scheduler to the workers, not from the client to the worker (right?). Serializing the graph piecewise is exactly what we do there.

**Comment by mrocklin** [comment]
Because the scheduler doesn't pickle or unpickle anything we serialize piecewise from the beginning at the client.  See `distributed/client.py::Client._graph_to_futures`.

File: dask/optimize.py [hunk 1]
```diff
@@ -611,3 +613,77 @@ def fuse_getitem(dsk, func, place):
...
+    """
+    constants = {}
+    key_lk = defaultdict(set)
+    for k, v in dsk.items():
+        for const in _extract_constants(v, predicate):
+            const_id = id(const)
+            constants[const_id] = const
+            key_lk[const_id].add(k)
+    key_lk = {c_id: k for c_id, k in key_lk.items() if len(k) > 1}
+    keys = set()
```
Comment (pitrou) [inline dask/optimize.py]: It would probably be a good idea to avoid recomputing `dsk2` if `key_lk` is empty.

Comment (jcrist) [inline dask/optimize.py]: Thanks, fixed.

**Comment by jcrist** [comment]
Should we merge this, or close it?

**Comment by mrocklin** [comment]
I no longer have an immediate need of this optimization in the wild.  I suggest that we leave it lingering as a closed PR until such a need arises.

**Comment by jcrist** [comment]
Fine by me.

**Pull request closed without merging**